src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,215 @@ namespace mlpack {
-  bootstrapLabels = labels.cols(indices);
-  if (UseWeights)
-    bootstrapWeights = weights.cols(indices);
-}
+    bootstrapDataset = dataset.cols(indices);
+    bootstrapLabels = labels.cols(indices);
+    if (UseWeights)
+      bootstrapWeights = weights.cols(indices);
+  }
+};
+
+/**
+ * No-op bootstrap strategy that just copies the input to the output.
+ */
+class IdentityBootstrap
+{
+ public:
+  template<bool UseWeights,
+           typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    MakeAlias(bootstrapDataset, dataset, dataset.n_rows, dataset.n_cols);
+    MakeAlias(bootstrapLabels, labels, labels.n_elem);
+    if (UseWeights)
+      MakeAlias(bootstrapWeights, weights, weights.n_elem);
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ *
+ * @tparam IndMatType Indicator matrix type.
+ */
+template<typename IndMatType = arma::umat>
+class SequentialBootstrap
+{
+ public:
+  /**
+   * Constructor.
+   *
+   * @param[in] intervals See ComputeAverageUniqueness().
+   */
+  SequentialBootstrap(const IndMatType& intervals) :
+    intervals(intervals)
+  {
+    if (intervals.n_cols != 2)
+      throw std::invalid_argument(
+        "SequentialBootstrap::SequentialBootstrap(): "
+        "intervals must be a 2 x m matrix!");
+  }
+
+  template<bool UseWeights,
+    typename MatType,
+    typename LabelsType,
+    typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+    const LabelsType& labels,
+    const WeightsType& weights,
+    MatType& bootstrapDataset,
+    LabelsType& bootstrapLabels,
+    WeightsType& bootstrapWeights)
+  {
+    if (labels.n_cols != intervals.n_rows)
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+        "labels n_cols and intervals n_rows differ!");
+
+    // observations are stored as columns and dimensions
+    // (number of features) as rows.
+    const arma::uvec phi(arma::conv_to<arma::uvec>::from(
+      ComputeSamples(dataset.n_cols)));
+
+    bootstrapDataset = dataset.cols(phi);
+    bootstrapLabels = labels.cols(phi);
+    if (UseWeights)
+      bootstrapWeights = weights.rows(phi);
+  }
+
+  /**
+   * Compute the samples of the next draw.
+   *
+   * @param[in] colCount Number of data points in the dataset.
+   *
+   * @return A list of indices referring to the observations that should
+   *         be sampled.
+   */
+  std::vector<arma::uword> ComputeSamples(arma::uword colCount) const
+  {
+    DiscreteDistribution     d;
+    std::vector<arma::uword> phi;
+
+    phi.reserve(colCount);
+    for (arma::uword i(0); i < colCount; ++i) {
+      d.Probabilities() = ComputeNextDrawProbabilities(phi, intervals);
+      phi.push_back(d.Random()[0]);
```

Comment:
If `d.Probabilities()` is set to the average uniqueness of each *interval*, don't we have to then sample from each interval?  So `d.Random()[0]` gets you a random interval (weighted by uniqueness), and then you need to randomly draw a sample in the range `(begin, end)` of that interval?

Reply:
`phi` holds the set of samples that we already sampled (drew again). In the beginning this set is empty. We draw until we have `N` samples (`N` can be the original amount of data points but because we draw with replacement and not requiring uniqueness as in the paper you referenced above, certain data points may appear multiple times in the end).

In every iteration we hypothetically add each data point to `phi` and calculate the average uniqueness for that data point given that some other data points are already in `phi`. We then use those average uniquenesses as probabilities to actually draw randomly with probabilities one of the data points, and so on and so forth